src/modules.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use utils;
+
+use std::path::{Path, PathBuf};
+use std::collections::HashMap;
+
+use syntax::ast;
+use syntax::codemap;
+use syntax::parse::{parser, token};
+
+
+/// List all the files containing modules of a crate.
+/// If a file is used twice in a crate, it appears only once.
+pub fn list_modules<'a>(krate: &'a ast::Crate,
+                        codemap: &codemap::CodeMap)
+                        -> HashMap<PathBuf, &'a ast::Mod> {
+    let mut result = HashMap::new();
+    let root_filename: PathBuf = codemap.span_to_filename(krate.span).into();
+    list_submodules(&krate.module, root_filename.parent().unwrap(), codemap, &mut result);
+    result.insert(root_filename, &krate.module);
+    result
+}
+
+/// Recursively list all external modules included in a module.
+fn list_submodules<'a>(module: &'a ast::Mod,
+                       search_dir: &Path,
+                       codemap: &codemap::CodeMap,
+                       result: &mut HashMap<PathBuf, &'a ast::Mod>) {
+    debug!("list_submodules: search_dir: {:?}", search_dir);
+    for item in module.items.iter() {
```

Comment:
We need access to the span, attrs and ident of the item inside the loop, and something like
```rust
let filter = |item| match item.node {
    ast::ItemMod(ref sub_mod) if !utils::contains_skip(&item.attrs) => Some((sub_mod, item.span, item.attrs, item.ident)),
    _ => None
};

for (sub_mod, span, attrs, ident) in module.items.iter().filter_map(filter) {
    let is_internal = ...
}
```
is IMHO less readable.
Another possibility is to put the match inside the loop, but it doesn't reduce rightwards drift.